[IDLE-240] 관심도서, 완독도서, 소장도서 사용자 조회 API 개발

minnseong opened the pull request:
## 🧑‍💻 작업 사항

### 관심 도서에 대해서 사용자 조회 API 개발
- GET "/api/v1/books/{isbn13}/interest/userInfos"
```
{
  "status": "200 OK",
  "message": "관심 도서를 등록한 유저 목록 조회에 성공했습니다.",
  "data": {
    "userInfos": [
      {
        "userId": "USER_ID_1",
        "nickname": "USER_NICKNAME_1",
        "profileImageUrl": "USER_PICTURE_URL_1"
      },
      {
        "userId": "USER_ID_2",
        "nickname": "USER_NICKNAME_2",
        "profileImageUrl": "USER_PICTURE_URL_2"
      }
    ]
  }
}
```

### 완독 도서에 대해서 사용자 조회 API 개발
- 완독 도서이며, 공개 마이북, 삭제되지 않은 마이북 설정인 경우
- GET "/api/v1/books/{isbn13}/read-complete/userInfos"
```
{
  "status": "200 OK",
  "message": "도서를 완독한 유저 목록 조회에 성공했습니다.",
  "data": {
    "userInfos": [
      {
        "userId": "USER_ID_1",
        "nickname": "USER_NICKNAME_1",
        "profileImageUrl": "USER_PICTURE_URL_1"
      },
      {
        "userId": "USER_ID_2",
        "nickname": "USER_NICKNAME_2",
        "profileImageUrl": "USER_PICTURE_URL_2"
      }
    ]
  }
}
```

### 소장 도서에 대해서 사용자 조회 API 개발
- 공개 마이북, 삭제되지 않은 마이북 설정인 경우
- GET "/api/v1/books/{isbn13}/mybook/userInfos"
```
{
  "status": "200 OK",
  "message": "도서를 마이북에 등록한 유저 목록 조회에 성공했습니다.",
  "data": {
    "userInfos": [
      {
        "userId": "USER_ID_1",
        "nickname": "USER_NICKNAME_1",
        "profileImageUrl": "USER_PICTURE_URL_1"
      },
      {
        "userId": "USER_ID_2",
        "nickname": "USER_NICKNAME_2",
        "profileImageUrl": "USER_PICTURE_URL_2"
      }
    ]
  }
}
```

### 유저 정보 호출 feignClient API 수정
- 기존: POST "/api/v1/users/info" 
    - @RequestBody를 사용하여 userId 전달
- 변경 GET "/api/v1/users/info?userId=AAA&userId=BBB&userId=CCC"
    - (@RequestParam("userId") List<String> userIds) 쿼리 파라미터에 리스트 사용하여 userId 전달
- 페이징 처리 도입 예정

### SonarCloud 자동화 도입
- master 브랜치 push 혹은 develop 브랜치 pull_request 시 동작
- 아래 지표 측정
<img width="181" alt="image" src="https://github.com/SWM-IDLE/mybrary-book-service/assets/71378475/014d89cb-4230-4480-8e80-9220b1d7408e">



<br><br>

## 🔗 링크

<!-- 관련된 대화가 이루어진 슬랙 링크 혹은 피그마 링크를 첨부. -->
<!-- - [프레이머](https://framer.com/projects/2--gSDcZoNKqU6dqCUQUYQe-53rEr?node=tQMWSSKqy)  -->

<br><br>

## 🐰 시급한 정도

<!-- 🏃‍♂️ 보통 : 최대한 빠르게 리뷰 부탁드립니다. -->
<!-- 🚨 긴급 : 선 어프루브 후 리뷰를 부탁드립니다. (리뷰는 추후 반영)  -->
<!-- 🐢 천천히 : 급하지 않습니다. -->

<br><br>

## 📖 참고 사항

<!-- 공유할 내용, 레퍼런스, 추가로 발생할 것으로 예상되는 이슈, 스크린샷 등을 넣어 주세요.  -->
<!-- 변경사항이 큰 경우 집중해야 할 부분, 또는 불안해서 봐주었으면 하는 부분 등 -->

minnseong committed:
[IDLE-240] feat: 유저 정보 호출 feignClient API 수정
- 기존 : POST HTTP 메서드로 @RequestBody에 UserIds를 담어서 호출
- 수정 : GET HTTP 메서드로 (@RequestParam("userId") List<String> userIds)를 통해 호출

minnseong committed:
[IDLE-240] refactor: MyBookService 분리
- MyBookService를 MyBookReadService와 MyBookWriteService를 분리

minnseong committed:
[IDLE-240] feat: 한 도서에 대해서 관심 도서로 설정한 유저 ID 조회 QueryDsl 작성
- 도서를 매개변수로 받아 해당 도서를 관심 도서로 설정한 유저들의 ID을 반환한다.

minnseong committed:
[IDLE-240] feat: 한 도서에 대해서 관심 도서로 설정한 유저 ID 조회 서비스 코드 작성
- UserServiceClient를 사용하여 feignClient를 통해 UserInfo 조회

minnseong committed:
[IDLE-240] feat: 한 도서에 대해서 관심 도서로 설정한 유저 ID 조회 API 구현
- GET "api/v1/books/{isbn13}/interest/userInfos"

minnseong committed:
[IDLE-240] feat: 한 도서를 완독한 유저의 목록 조회 QueryDsl 작성
- Book을 매개변수로 받아, 완독되었고, 공개이고, 삭제되지 않는 MyBook의 userId를 조회한다.
- queryDsl을 사용하여 userId만 조회 가능하도록 한다.

minnseong committed:
[IDLE-240] feat: 한 도서에 대해서 완독한 유저 정보 조회 서비스 코드 작성
- UserServiceClient를 사용하여 feignClient를 통해 UserInfo 조회

minnseong committed:
[IDLE-240] feat: 한 도서에 대해서 완독한 유저 정보 조회 API 구현
- GET "/api/v1/books/{isbn13}/read-complete/userInfos"

minnseong committed:
[IDLE-240] feat: 한 도서를 마이북으로 설정한 유저의 목록 조회 QueryDsl 작성
- Book을 매개변수로 받아, 완독되었고, 공개이고, 삭제되지 않는 MyBook의 userId를 조회한다.
- queryDsl을 사용하여 userId만 조회 가능하도록 한다.

minnseong committed:
[IDLE-240] feat: 한 도서에 대해서 마이북으로 설정한 유저 정보 조회 서비스 코드 작성
- UserServiceClient를 사용하여 feignClient를 통해 UserInfo 조회

minnseong committed:
[IDLE-240] feat: 한 도서에 대해서 마이북으로 등록한 유저 정보 조회 API 구현
- GET "/api/v1/books/{isbn13}/mybook/userInfos"

minnseong committed:
[IDLE-240] refactor: 유저 정보 조회 메서드명 변경
- getReadCompletedUserIdListByBook -> getUserIdWithReadCompletedListByBook
- getMyBookSetUserIdListByBook -> getUserIdListWithMyBookSettingByBook

minnseong committed:
[IDLE-240] cicd: github workflow 수정
- ecr-ecs 자동화는 master 브랜치에 push 되는 경우 수행
- rest-docs 자동화는 develop 브랜치에 pr 되는 경우 수행

minnseong committed:
[IDLE-240] cicd: SonarCloud 자동화 workflow 작성

minnseong committed:
[IDLE-240] cicd: SonarCloud 자동화 workflow 수정

minnseong committed:
[IDLE-240] cicd: jacoco 자동화 workflow 작성

minnseong committed:
[IDLE-240] cicd: jacoco 자동화 workflow 작성

minnseong committed:
[IDLE-240] cicd: jacoco 자동화 workflow 수정

minnseong committed:
[IDLE-240] cicd: jacoco 자동화 workflow 수정

minnseong committed:
[IDLE-240] cicd: jacoco 자동화 workflow 수정

minnseong committed:
[IDLE-240] cicd: jacoco 자동화 workflow 수정

minnseong committed:
[IDLE-240] refactor: SonarCloud Code Smell 수정
- 테스트 코드 hasSize() 사용
- ResponseEntity 타입 선언

minnseong committed:
[IDLE-240] refactor: SonarCloud Code Smell 수정
- Provide the parametrized type for this generic Rule 해결
- ResponseEntity -> ResponseEntity<SuccessResponse<UserInfoWithMyBookSetForBookResponse>>

0112leesy commented on book-service/src/main/java/kr/mybrary/bookservice/client/user/api/UserServiceClient.java:
```diff
@@ -3,34 +3,34 @@
     @Retry(name = "userServiceRetryConfig", fallbackMethod = "getUsersInfoFallback")
     @CircuitBreaker(name = "userServiceCircuitBreakerConfig", fallbackMethod = "getUsersInfoFallback")
-    UserInfoServiceResponse getUsersInfo(@RequestBody UserInfoRequest userInfoRequest);
+    UserInfoServiceResponse getUsersInfo(@RequestParam("userId") List<String> userIds);
```
q: 혹시 RequestParam 요청 형태가 어떻게 되는지 예시를 주실 수 있을까용?

minnseong replied:
위와 같이 feignClient로 요청을 보내면, 아래와 같이 요청을 보낼 수 있습니다.
```
GET /api/v1/users/info?userId=user1&userId=user2&userId=user3
```

따라서 user-service 에서 아래와 같이 API를 작성하시면 될 것 같습니다.
```
@GetMapping("/api/v1/users/info")
public ResponseEntity getUsersInfo(@RequestParam("userId") List<String> userIds) {
 
    ...
}
```

0112leesy commented on book-service/src/main/java/kr/mybrary/bookservice/mybook/domain/dto/request/UserInfoWithMyBookSetForBookServiceRequest.java:
```diff
@@ -0,0 +1,18 @@
+
+@Getter
+@Builder
+public class UserInfoWithMyBookSetForBookServiceRequest {
```
q: DTO명이 길어져서 UserInfoWithMyBookSettingForBookServiceRequest 대신 UserInfoWithMyBookSetForBookServiceRequest로 의도하신게 맞나요?

minnseong replied:
네! 너무 DTO 명이 길어서 변경했지만, 일관성을 맞추는게 좀 더 괜찮을 것 같네요..!
일관성 맞도록 수정해보겠습니다!

0112leesy reviewed (approved):
신속하게 API 개발하시느라 고생많으셨습니다! SonarCloud를 추가하니 확실히 코드 품질이 관리되고 있다는 느낌이 드네요 😄

minnseong committed:
[IDLE-240] refactor: DTO 명 변경
- 한 도서에 대해서 마이북으로 설정한 유저 정보에 관한 DTO명이 UserInfoWithMyBookSetForBookResponse와 같이 Set 사용
- 하지만 이를 사용하는 메서드명에는 Set 대신 Setting을 사용하였기에 일관성을 맞추기 위해 Set 대신 Setting으로 수정
